### PR TITLE
DNSProxy: Provide a flag for rejected DNS queries.

### PR DESCRIPTION
**Pull request by eloycoto**
Due Alpine Linux uses musl libc library the dns support is not the same
as libc. One of the biggest different is that Alpine Linux containers
under Kubernetes will query using the search string from
resolv.conf, like `www.google.com.default.svc.cluster.local.`

The default response code is Refused(5) but in this case Alpine Linux
will not do the query again with the correct fqdn. So this flag allows
Cilium to reply NXDomain flag(3) so the dns flow will work as normal.

The new flag is `--tofqdns-dns-response-code=3`

The list of valid responses codes are:
https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6571)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

File: daemon/daemon_main.go [hunk 1]
```diff
@@ -525,6 +527,8 @@ func init() {
 	flags.BoolVar(&enableDNSPoller,
 		"tofqdns-enable-poller", false, "Enable proactive polling of DNS names in toFQDNs.matchName rules.")
 
+	flags.IntVar(&rejectDNSResponseCode, "tofqdns-dns-response-code", dnsproxy.ProxyDenyWithRefused, "DNS response code for rejecting DNS request")
```
Comment (tgraf) [inline daemon/daemon_main.go]: I think it would be a lot more user friendly to allow specifying this as string and do the translation to the opcode in `SetRejectReply()`. While you are at it, please define the option name in `pkg/option` and add a `BindEnv()` so users can use the ConfigMap easily.

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by eloycoto** [comment]
test-me-please

File: pkg/fqdn/dnsproxy/proxy.go [hunk 2]
```diff
@@ -44,6 +44,20 @@ const (
...
+	ProxyDenyWithNameError = "nameError"
+
+	// ProxyDenyWithRefused is the response code for Domain refused. It is the
+	// default for denied DNS requests.
+	ProxyDenyWithRefused = "refused"
+)
+
+var (
+	rejectReply   = dns.RcodeRefused
+	RejectOptions = []string{ProxyDenyWithNameError, ProxyDenyWithRefused}
```
[the remark this answers is not in the document]
Comment (vadorovsky) [reply pkg/fqdn/dnsproxy/proxy.go]: Shouldn't this variable be in `options` package? That's where we usually store default values.

File: pkg/fqdn/dnsproxy/proxy.go [hunk 3]
```diff
@@ -431,12 +445,32 @@ func bindToAddr(address string, port uint16) (UDPConn *net.UDPConn, TCPListener
...
 	if err = w.WriteMsg(refused); err != nil {
 		scopedLog.WithError(err).Error("Cannot send REFUSED response")
 		err = fmt.Errorf("cannot send REFUSED response: %s", err)
 	}
 	return err
 }
+
+// SetRejectReply sets the default reject reply on denied dns responses.
+func SetRejectReply(option string) {
+	switch option {
```
Comment (tgraf) [inline pkg/fqdn/dnsproxy/proxy.go]: I would add a `strings.ToLower()` here

**Review by tgraf — CHANGES_REQUESTED** [review]
Almost perfect. I agree with the suggestion from @mrostecki

**Comment by eloycoto** [comment]
test-me-please

File: pkg/fqdn/dnsproxy/proxy.go [hunk 4]
```diff
@@ -46,6 +47,10 @@ const (
 	ProxyBindRetryInterval = ProxyBindTimeout / 5
 )
 
+var (
+	rejectReply = dns.RcodeRefused
```
Comment (raybejjani) [inline pkg/fqdn/dnsproxy/proxy.go]: I have a slight preference for making this a property on the proxy object itself. It is easily changeable, though, so it's not a big deal.

File: pkg/fqdn/dnsproxy/proxy.go [hunk 5]
```diff
@@ -431,12 +436,32 @@ func bindToAddr(address string, port uint16) (UDPConn *net.UDPConn, TCPListener
...
+		rejectReply = dns.RcodeRefused
+	default:
+		log.Infof("DNS reject response '%d' is not valid, available options are '%v'",
+			opt, option.FQDNRejectOptions)
+		return
+	}
+}
+
+// getRegectReply returns a DNS message with the correct response code.
+func getRejectReply(request *dns.Msg) *dns.Msg {
```
Comment (raybejjani) [inline pkg/fqdn/dnsproxy/proxy.go]: generateRejectReply would be a little better. When I first read get/set I thought we had 1 reply that we re-used.

File: pkg/option/config.go [hunk 6]
```diff
@@ -328,7 +328,28 @@ const (
...
+	MonitorQueueSizeNameEnv = "CILIUM_MONITOR_QUEUE_SIZE"
+
+	//FQDNRejectResponseCode is the name for the option for dns-proxy reject response code
+	FQDNRejectResponseCode = "tofqdns-dns-response-code"
+
+	//FQDNRejectResponseCodeEnv is the env name for FQDNRejectResponseCode option
+	FQDNRejectResponseCodeEnv = "CILIUM_TOFQDNS_DNS_RESPONSE_CODE"
+
+	// FQDNProxyDenyWithNameError is the response code for Domain does not
+	// exists.  It is only used when musl libc library is in place (Alpine
```
Comment (raybejjani) [inline pkg/option/config.go]: mentioning musl is ok, but it isn't the only case, just the one we found. Something like: It is useful when stub resolvers, like the one in Alpine Linux's libc (musl), treat a REFUSED as a resolution error. This happens when trying a DNS search list, as in kubernetes, and breaks even whitelisted DNS names.

File: daemon/daemon_main.go [hunk 7]
```diff
@@ -600,6 +600,9 @@ func init() {
 	flags.Bool(option.ToFQDNsEnablePoller, false, "Enable proactive polling of DNS names in toFQDNs.matchName rules.")
 	option.BindEnv(option.ToFQDNsEnablePoller)
 
+	flags.StringVar(&option.Config.FQDNRejectResponse, option.FQDNRejectResponseCode, option.FQDNProxyDenyWithRefused, fmt.Sprintf("DNS response code for rejecting DNS request, available options are '%v'", option.FQDNRejectOptions))
```
Comment (raybejjani) [inline daemon/daemon_main.go]: Could you change this to "for rejecting DNS requests" (the plural part)?

File: pkg/option/config.go [hunk 8]
```diff
@@ -328,7 +328,28 @@ const (
 	EnableIPv6Name = "enable-ipv6"
 
 	// MonitorQueueSizeName is the name of the option MonitorQueueSize
-	MonitorQueueSizeName = "monitor-queue-size"
+	MonitorQueueSizeName    = "monitor-queue-size"
+	MonitorQueueSizeNameEnv = "CILIUM_MONITOR_QUEUE_SIZE"
+
+	//FQDNRejectResponseCode is the name for the option for dns-proxy reject response code
+	FQDNRejectResponseCode = "tofqdns-dns-response-code"
```
Comment (raybejjani) [inline pkg/option/config.go]: Although longer and uglier, I wonder if `tofqdns-dns-reject-response-code` is a better name. I'm worried we will one day end up needing to handle some other situation. I could be convinced otherwise, though, because I can't imagine this situation :/

File: daemon/daemon_main.go (hunk 7 above)
Comment (raybejjani) [inline daemon/daemon_main.go]: Is there a way to make the stringvar case-insensitive? then we would worry less about nameError vs NameError

File: Documentation/cmdref/cilium-bugtool.md [hunk 9]
```diff
@@ -6,10 +6,11 @@ Collects agent & system information useful for bug reporting
 
 ### Synopsis
 
+
 Collects agent & system information useful for bug reporting
 
 ```
-cilium-bugtool [OPTIONS] [flags]
```
Comment (raybejjani) [inline Documentation/cmdref/cilium-bugtool.md]: Could you put this file change in a separate commit? or even just drop it from the PR. I'm not sure what changed to make the help pop up, a different version of the generator?

File: daemon/daemon.go [hunk 10]
```diff
@@ -1389,6 +1389,7 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState) (err err
 	if option.Config.ToFQDNsEnablePoller {
 		fqdn.StartDNSPoller(d.dnsPoller)
 	}
+	dnsproxy.SetRejectReply(option.Config.FQDNRejectResponse)
```
Comment (raybejjani) [inline daemon/daemon.go]: Since this is a configuration thing, could you put it just before we init the proxy? Alternatively, it can go with cfg above.

**Review by raybejjani — APPROVED** [review]
Nothing blocking, but a bunch of small reshuffles or edits.

File: daemon/daemon_main.go (hunk 7 above)
Comment (eloycoto) [reply daemon/daemon_main.go]: all used as ToLower now

File: pkg/fqdn/dnsproxy/proxy.go (hunk 4 above)
Comment (eloycoto) [reply pkg/fqdn/dnsproxy/proxy.go]: The thing, if I change from proxy I need to change the Refused function to be part of the proxy. I'm happy to do that, but want to make sure.

**Comment by eloycoto** [comment]
test-me-please

File: pkg/fqdn/dnsproxy/proxy.go (hunk 4 above)
Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy.go]: Yeah, I also meant that. It could be a helper that take the RCode and the request message, if you felt the hidden state was a little 0_o I'm not too sure that this matters overall... and we can change it when we need to, so feel free to not make it.

File: Documentation/cmdref/cilium-bugtool.md (hunk 9 above)
Comment (eloycoto) [reply Documentation/cmdref/cilium-bugtool.md]: I couldn't not deleted, if not the CMDRef fail :s It's all rebased

File: pkg/fqdn/dnsproxy/proxy.go (hunk 4 above)
Comment (eloycoto) [reply pkg/fqdn/dnsproxy/proxy.go]: made this comment, If you're agree I'll squash the commits: 
https://github.com/cilium/cilium/pull/6571/commits/3179e872b6db1af4ee61766776ae67d05099224e

Comment (raybejjani) [reply pkg/fqdn/dnsproxy/proxy.go]: Perfect. Thank you!

**Comment by eloycoto** [comment]
test-me-please (cmd-ref issues with bugtool) https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8934/execution/node/92/log/

**Review by tgraf — APPROVED** [review]

**Comment by eloycoto** [comment]
test-me-please
